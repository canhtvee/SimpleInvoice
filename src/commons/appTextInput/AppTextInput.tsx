import React, {useState} from 'react';
import {
  StyleProp,
  StyleSheet,
  TextInput,
  TextInputProps,
  TextStyle,
  View,
  ViewStyle,
} from 'react-native';
import {Controller, UseControllerProps, useFormState} from 'react-hook-form';
import Feather from 'react-native-vector-icons/Feather';

import {Colors, Sizes} from '../../utils';

import {AppText} from '../appText';

export interface AppTextInputProps
  extends UseControllerProps,
    Omit<TextInputProps, 'defaultValue'> {
  label?: string;
  labelStyle?: StyleProp<TextStyle>;
  errorStyle?: StyleProp<TextStyle>;
  containerStyle?: StyleProp<ViewStyle>;
  inputContainerStyle?: StyleProp<ViewStyle>;
  inputStyle?: StyleProp<TextStyle>;
  leftChild?: JSX.Element;
  rightChild?: JSX.Element;
}

export function AppTextInput({
  control,
  name,
  label,
  labelStyle,
  errorStyle,
  defaultValue = '',
  rules,
  secureTextEntry,
  inputStyle,
  inputContainerStyle,
  containerStyle,
  maxLength,
  ...textInputProps
}: AppTextInputProps) {
  const {errors} = useFormState({control, name});
  const [secure, setSecure] = useState(secureTextEntry);

  return (
    <View style={containerStyle}>
      {label && <AppText style={labelStyle}>{label}</AppText>}
      <View style={[styles.inputContainer, inputContainerStyle]}>
        <Controller
          defaultValue={defaultValue}
          name={name}
          control={control}
          rules={rules}
          render={({field: {onChange, value, onBlur, ref}}) => (
            <TextInput
              onBlur={onBlur}
              ref={ref}
              onChange={onChange}
              value={value}
              maxLength={rules?.maxLength?.value || maxLength}
              autoCapitalize={'none'}
              autoCorrect={false}
              spellCheck={false}
              style={[
                styles.input,
                {
                  color: Colors.text,
                },
                inputStyle,
              ]}
              secureTextEntry={secure}
              {...textInputProps}
            />
          )}
        />

        {secureTextEntry && (
          <Feather
            name={secure ? 'eyes' : 'eyesOff'}
            style={{paddingRight: Sizes.paddinglx}}
            onPress={() => setSecure(prev => !prev)}
          />
        )}
      </View>

      {errors[name] && errors[name]?.message && (
        <AppText
          style={[
            {
              color: Colors.error,
              marginTop: Sizes.paddinglx,
            },
            errorStyle,
          ]}>
          {errors[name]?.message as string}
        </AppText>
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  inputContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    overflow: 'hidden',
    borderWidth: Sizes.borderWidth,
    borderRadius: Sizes.borderRadius,
  },
  input: {
    flex: 1,
    fontSize: Sizes.regular,
    paddingHorizontal: Sizes.paddinglx,
  },
});
