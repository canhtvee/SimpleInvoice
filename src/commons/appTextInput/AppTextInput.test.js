import React from 'react';
import {useForm} from 'react-hook-form';
import {cleanup, renderHook} from '@testing-library/react-native';
import {create} from 'react-test-renderer';

import {AppTextInput} from './AppTextInput';

/**
 * Run test command
 * yarn test -u -t="AppTextInput"
 */

afterEach(cleanup);

describe('AppTextInput', () => {
  test('should render correctly', () => {
    const {result} = renderHook(() => {
      return useForm();
    });

    const {control} = result.current;

    const tree = create(
      <AppTextInput
        control={control}
        name={'test-input'}
        label={'test-label'}
        placeholder={'test-input'}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();

    expect(1).toBeDefined();
  });
});
