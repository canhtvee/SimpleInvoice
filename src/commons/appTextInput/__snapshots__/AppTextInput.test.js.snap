// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppTextInput should render correctly 1`] = `
<View>
  <Text
    style={
      [
        {
          "color": "#000000",
          "fontSize": 27.5,
        },
        undefined,
      ]
    }
  >
    test-label
  </Text>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderColor": "#0000008A",
          "borderRadius": 6,
          "borderWidth": 1,
          "flexDirection": "row",
          "overflow": "hidden",
        },
        undefined,
      ]
    }
  >
    <TextInput
      autoCapitalize="none"
      autoCorrect={false}
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="test-input"
      placeholderTextColor="#00000061"
      spellCheck={false}
      style={
        [
          {
            "flex": 1,
            "fontSize": 27.5,
            "paddingHorizontal": 16,
          },
          {
            "paddingHorizontal": 16,
            "paddingVertical": 23.5,
          },
          {
            "color": "#000000",
          },
          undefined,
        ]
      }
      value=""
    />
  </View>
</View>
`;
